Seun save state

ogewan opened the pull request:
Minimum Viable Product

nkemnitz commented on src/neuroglancer/save_state/save_state.ts:
```diff
@@ -0,0 +1,333 @@
+  constructor(public root: Trackable, updateDelayMilliseconds = 400) {
+    super();
+    const saverManuallyDisabled = !getOldStyleSaving().value;
+    this.supported = !storageAvailable() && saverManuallyDisabled;
```
This seems confusing - state saving is supported when there is no storage available and the user manually disabled state saving? Either the check or the variable names don't match...

ogewan commented:
Ready for Review

ogewan commented:
Significant changes since last review:
sid changed to local_id
local_id corresponds to a tabbed session and no longer changes
Individual states are no longer saved, it is grouped into the tabbed session
save button changed to share
sharing state provides json link and stores in history if state server is accessible
sharing a state that has not been updated since last share will provide the same links provided previously without pushing to state server.

keyboard shortcuts:
ctrl + shift + s = save and save window
ctrl + shift + j = save and copy json link or raw link if no json link
ctrl + shift + r = save and copy raw link

nkemnitz commented on src/neuroglancer/viewer.ts:
```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
+          .finally(() => {
+            const noStateServerAccess = !this.jsonStateServer.value;
+            if (noStateServerAccess) {
+              this.hashBinding!.legacy.setUrlHash();
```
`hashBinding` _could_ be undefined, right? I thought it's only initialized if local storage does not work?

ogewan replied:
added a comment explaining. hashBinding is not optional hashBinding.legacy.falback is

nkemnitz commented on src/neuroglancer/viewer.ts:
```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
     }
+    else {
+      StatusMessage.showTemporaryMessage(`Cannot access state server. Press the share button/CTRL + SHIFT + S to enter a server URL.`);
+      this.showSaveDialog(2);
```
Use `UrlType.raw`

ogewan commented on src/neuroglancer/viewer.ts:
```diff
@@ -202,9 +207,16 @@ function makeViewerContextMenu(viewer: Viewer) {
 export class Viewer extends RefCounted implements ViewerState {
   navigationState = this.registerDisposer(new NavigationState());
   perspectiveNavigationState = new NavigationState(new Pose(this.navigationState.position), 1);
+  saver?: SaveState;
```
Unfortunately, it depends on loadFromJson which is async and cannot be done in the constructor. But I moved initialization of setup into Viewer

ogewan commented on src/neuroglancer/viewer.ts:
```diff
@@ -202,9 +207,16 @@ function makeViewerContextMenu(viewer: Viewer) {
   navigationState = this.registerDisposer(new NavigationState());
   perspectiveNavigationState = new NavigationState(new Pose(this.navigationState.position), 1);
+  saver?: SaveState;
+  hashBinding?: UrlHashBinding;
```
Similarly, optional hashbinding depends on saver which cannot be constructed till the async load from json; moved to the same place as saver initialization.

ogewan commented on src/neuroglancer/viewer.ts:
```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
     }
+  }
 
+  postJsonState(get: UrlType = 0) {
```
UrlType = 0 is equivalent to not setting the UrlType which is the case for CTRL + SHIFT + S and pressing the share button normal.

ogewan replied:
changed UrlType to a conditional parameter

ogewan commented on src/neuroglancer/preferences/user_preferences.ts:
```diff
@@ -10,19 +10,19 @@ class UserPreferences {
   prefetchSliceViewChunks: TrackableBoolean;
   cursorOnMousedrag: TrackableBoolean;
-  urlAutoSave: TrackableBoolean;
+  oldStyleSave: TrackableBoolean;
```
changed to saveToAddressBar

nkemnitz commented on src/neuroglancer/save_state/save_state.ts:
```diff
@@ -0,0 +1,383 @@
+      jsonUrl = `${urlStart}?json_url=${jsonString}`;
+    } else {
+      if (!jsonUrl) {
+        jsonUrl = 'NOT AVALABLE';
```
Typo: "available"
Maybe make it a constant to ensure that we are calling it the same everywhere in the code.

nkemnitz commented on src/neuroglancer/save_state/save_state.ts:
```diff
@@ -0,0 +1,383 @@
+
+    form.append(this.makePopup('JSON_URL'));
+    this.insertField(
+        form, 'JSON_URL', jsonUrl, 'neuroglancer-save-state-json', jsonUrl === 'NOT AVALABLE');
```
Typo: "available", see earlier comment.

nkemnitz commented on src/neuroglancer/save_state/save_state.ts:
```diff
@@ -0,0 +1,383 @@
+    super();
+
+    let urlStart = `${window.location.origin}${window.location.pathname}`;
+    let jsonUrl;
```
Can just initialize `jsonURL` here and remove the `else` branch below.

nkemnitz commented on src/neuroglancer/save_state/save_state.ts:
```diff
@@ -0,0 +1,383 @@
+}
+
+class SaveDialog extends Overlay {
+  constructor(public viewer: Viewer, jsonString?: string, get?: UrlType) {
```
`get` should be a bit more descriptive, e.g. `getUrlType`

nkemnitz commented on src/neuroglancer/viewer.ts:
```diff
@@ -742,27 +790,18 @@ export class Viewer extends RefCounted implements ViewerState {
     new UserReportDialog(this, image);
   }
 
+  showSaveDialog(get?: UrlType, jsonString?: string) {
```
Copied from `SaveDialog`: `get` should be a bit more descriptive, e.g. `getUrlType`

nkemnitz commented on src/neuroglancer/viewer.ts:
```diff
@@ -773,31 +812,70 @@ export class Viewer extends RefCounted implements ViewerState {
     }
+  }
 
+  postJsonState(get?: UrlType) {
```
Copied from `SaveDialog`: `get` should be a bit more descriptive, e.g. `getUrlType`
